drivers/gpio/gpio_grgpio2.c

```diff
@@ -0,0 +1,294 @@
+
+	regs->output_or = pins;
+	return 0;
+}
+
+static int port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct cfg *cfg = dev->config;
+	volatile struct grgpio_regs *regs = cfg->regs;
+
+	regs->output_and = ~pins;
+	return 0;
+}
+
+static int port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct cfg *cfg = dev->config;
+	volatile struct grgpio_regs *regs = cfg->regs;
+
+	regs->output_xor = pins;
+	return 0;
+}
+
+static uint32_t get_pending_int(const struct device *dev)
+{
+	const struct cfg *cfg = dev->config;
+	volatile struct grgpio_regs *regs = cfg->regs;
+
+	return regs->iflag;
+}
+
+static int pin_interrupt_configure(const struct device *dev,
+					      gpio_pin_t pin,
+					      enum gpio_int_mode mode,
+					      enum gpio_int_trig trig)
+{
+	const struct cfg *cfg = dev->config;
+	struct data *data = dev->data;
+	volatile struct grgpio_regs *regs = cfg->regs;
+	int ret = 0;
+	const uint32_t mask = 1 << pin;
+	uint32_t polmask;
+	k_spinlock_key_t key;
+
+	if ((mask & data->imask) == 0) {
+		/* This pin can not generate interrupt */
+		return -ENOTSUP;
+	}
+	if (mode != GPIO_INT_MODE_DISABLED) {
+		if (trig == GPIO_INT_TRIG_LOW) {
+			polmask = 0;
+		} else if (trig == GPIO_INT_TRIG_HIGH) {
+			polmask = mask;
+		} else {
+			return -ENOTSUP;
+		}
+	}
+	key = k_spin_lock(&data->lock);
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		regs->imask_and = ~mask;
+	} else if (mode == GPIO_INT_MODE_LEVEL) {
+		regs->imask_and = ~mask;
+		regs->iedge &= ~mask;
+		regs->ipol = (regs->ipol & ~mask) | polmask;
+		regs->imask_or = mask;
+	} else if (mode == GPIO_INT_MODE_EDGE) {
+		regs->imask_and = ~mask;
+		regs->iedge |= mask;
+		regs->ipol = (regs->ipol & ~mask) | polmask;
+		regs->imask_or = mask;
+	} else {
+		ret = -ENOTSUP;
+	}
+	k_spin_unlock(&data->lock, key);
+
+	int interrupt = cfg->interrupt;
+	const int irqgen = data->irqgen;
+
+	if (irqgen == 0) {
+		interrupt += pin;
+	} else if (irqgen == 1) {
+		;
+	} else if (irqgen < 32) {
+		/* look up interrupt number in GRGPIO interrupt map */
+		uint32_t val = regs->irqmap[pin/4];
+
+		val >>= (3 - pin % 4) * 8;
+		interrupt += (val & 0x1f);
+	}
+
+	if (interrupt && ((1 << interrupt) & data->connected) == 0) {
+		irq_connect_dynamic(
+			interrupt,
+			0,
+			(void (*)(const void *)) grgpio_isr,
+			dev,
+			0
+		);
+		irq_enable(interrupt);
+		data->connected |= 1 << interrupt;
+	}
+
+	return ret;
+}
+
+static int manage_callback(const struct device *dev,
+			   struct gpio_callback *callback,
+			   bool set)
+{
+	struct data *data = dev->data;
+
+	return gpio_manage_callback(&data->cb, callback, set);
+}
+
+static void grgpio_isr(const struct device *dev)
+{
+	const struct cfg *cfg = dev->config;
+	struct data *data = dev->data;
+	volatile struct grgpio_regs *regs = cfg->regs;
+	uint32_t pins;
+
+	/* no locking needed when iflag is implemented */
+	pins = regs->iflag;
+	if (pins == 0) {
+		return;
+	}
```

Comment:
I believe this check should be an assert, or we can remove it completely. If the driver is bug free it's not possible to have an interrupt without an interrupt source?

Reply:
It is possible that the ISR gets called when `regs->iflag == 0`.

The `iflag` is a status indicator local to the GPIO controller. A bit gets set to `1` in `iflag` when the corresponding I/O bit causes an interrupt condition. That condition is also forwarded to the interrupt controller as a request to set one of 31 pending bits.

A bit in `iflag` can be cleared by the user by writing `1`.  However, that write only affects the GPIO status indicator and does not acknowledge pending bits in the interrupt controller. The driver can clear bits in `iflag` which correspond to interrupts which are still pending in the interrupt controller. In that scenario, the ISR may get called after the driver has already cleared the `iflag` indication.

Acknowledgement of a pending interrupt in the LEON interrupt controller is performed automatically by hardware, at the same moment as the CPU starts executing the first instruction of the interrupt exception handler.